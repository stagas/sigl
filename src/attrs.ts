import type { Keys, StringOf, ValueConstructor } from 'everyday-types'
import { accessors } from 'everyday-utils'

/** @module attrs */

export type Attrs<T> = {
  [K in keyof T]: T[K] extends ValueConstructor ? ReturnType<T[K]> : T[K]
}

export const AttrTypes = new Map<ValueConstructor, (x: any) => string | number | boolean>([
  [String, (x: any): string => x.toString()],
  [Number, (x: any): number => parseFloat(x)],
  [Boolean, (x: any): boolean => (x = x === false ? false : x != null)],
])

export const applyAttrs = <T extends object>(
  self: any,
  data: Record<StringOf<keyof T>, T[StringOf<keyof T>] | void>,
  initialUpdates: (() => void)[],
) => {
  for (const [k, v] of Object.entries(data) as [StringOf<keyof T>, any][]) {
    //!? 'setting attr "%s"', k, v.toString()
    if ([String, Number, Boolean].includes(v)) self[k] = data[k] = void 0
    else if (v === true) (data as any)[k] = true
    else if (typeof v === 'object') data[k] = v.toString()
    //!? 'written attr:', data[k]
  }

  const update = (attr: any, key: any, value: any) => {
    if (typeof value === 'boolean') {
      if (value) {
        self.setAttribute(attr, '')
      } else {
        if (self.hasAttribute(attr)) {
          self.removeAttribute(attr)
        } else {
          // if the attribute is initially in the class as = true
          // but the actual html element doesn't have the attribute
          // then the removeAttribute() will not trigger
          // the attributeChangedCallback(), and the value will
          // stay at 'true' forever.
          ;(data as any)[key] = false
        }
      }
    } else {
      //!? 'setting', attr, value
      self.setAttribute(attr, value)
    }
  }

  return accessors(self, data, <K extends Keys<T>>(key: StringOf<K>) => {
    const attr = key.toLowerCase()
    return {
      get: () => data[key],
      set(value: T[K]) {
        //!warn 'setter changed', key, value
        if (!self.isConnected) initialUpdates.push(() => update(attr, key, value))
        else update(attr, key, value)
      },
    }
  })
}

export const attrListener = (body: string) =>
  new Function(
    'event',
    `with(this){let fn=${body};return typeof fn=='function'?fn.call(this,event):fn}`
  )
