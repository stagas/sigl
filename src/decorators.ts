import { toFluent } from 'to-fluent'

export class PropertySettings {
  attr = false
  out = false
  is?: PropertyKind
}

export type PropertyMap = Map<string, PropertySettings>

export interface PropertyKind<T = any> {
  from(x: unknown): T
  toString(x: T): string
}

export const protoPropertyMap = new WeakMap<object, PropertyMap>()

export const prop = toFluent(
  PropertySettings,
  settings =>
    (is?: PropertyKind): PropertyDecorator =>
      (proto, key) => {
        if (typeof key !== 'string') return
        settings.is ??= is
        if (!protoPropertyMap.has(proto)) {
          protoPropertyMap.set(proto, new Map([[key, settings]]))
        } else {
          const propertyMap = protoPropertyMap.get(proto)!
          if (propertyMap.has(key)) {
            Object.assign(propertyMap.get(key)!, settings)
          } else {
            propertyMap.set(key, settings)
          }
        }
      }
)

export const attr = prop.attr
export const out = prop.out
export const is = prop.is
