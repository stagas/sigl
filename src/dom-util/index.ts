export * from 'get-element-offset'
export { isMobileAgent as isMobile } from 'is-mobile-agent'
export * from './classes'
export * from './dataset'
export * from './find-parent-element'
export * from './get-active-element'
export * from './get-offset-rect'
export * from './get-observed-rect'
export * from './get-viewport-rect'
export * from './is-mac'
export * from './kbd'
export * from './observe'
export * from './ondblclick'
export * from './shadow'
export * from './transition'
