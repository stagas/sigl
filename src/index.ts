// @ts-ignore
globalThis.Reflect.metadata ??= () => {}

import { EventKeys } from 'everyday-types'
import * as sigl from './sigl'

export * from 'everyday-types'

const $ = Object.assign(sigl._, sigl, {
  String: String as unknown as string | undefined,
  Number: Number as unknown as number | undefined,
  Boolean: Boolean as unknown as boolean | undefined,
})

// This is painful but it's the only way i've found in order to
// merge the function $ with the types and methods from the exports.
// It needs to be updated with new exported types from the sigl namespace.
// TODO: do this programatically
namespace $ {
  export type AbortOptions = sigl.AbortOptions
  export type ChildOf<T> = sigl.ChildOf<T>
  export type Class<T> = sigl.Class<T>
  export type CleanClass<T> = sigl.CleanClass<T>
  export type CleanInstance<T> = sigl.CleanInstance<T>
  export type Component<T = HTMLElement, I = HTMLElement> = sigl.Component<T, I>
  export type Context<T extends object> = sigl.Context<T>
  export type ContextClass<T extends object> = sigl.ContextClass<T>
  export type CustomElementConstructor = sigl.CustomElementConstructor
  export type Deps<T> = sigl.Deps<T>
  export type Dispatch<T> = sigl.Dispatch<T>
  export type DispatchOptions = sigl.DispatchOptions
  export type Element<T, E = object> = sigl.Element<T, E>
  export type ElementClass = sigl.ElementClass
  export type EventHandler<T, E> = sigl.EventHandler<T, E>
  export type EventOptions = sigl.EventOptions
  export type Events<T, P extends Record<string, Event>> = sigl.Events<T, P>
  export type FxFn<T, R> = sigl.FxFn<T, R>
  export type HTMLAttributes<T> = sigl.HTMLAttributes<T>
  export type Intersect = sigl.Intersect
  export type JsxContext<T> = sigl.JsxContext<T>
  export type LifecycleEvents = sigl.LifecycleEvents
  export type Line = sigl.Line
  export type Matrix = sigl.Matrix
  export type Mixin = sigl.Mixin
  export type Morph = sigl.Morph
  export type Narrow<T, U> = sigl.Narrow<T, U>
  export type NestedCSSCompiler = sigl.NestedCSSCompiler
  export type NestedCSSDeclaration = sigl.NestedCSSDeclaration
  export type Off = sigl.Off
  export type On<T> = sigl.On<T>
  export type OnEvent<T, K extends EventKeys<T>> = sigl.OnEvent<T, K>
  export type OnGetter<T> = sigl.OnGetter<T>
  export type Point = sigl.Point
  export type Polygon = sigl.Polygon
  export type Polyline = sigl.Polyline
  export type PropertyKind<T> = sigl.PropertyKind<T>
  export type QueueOptions = sigl.QueueOptions
  export type Rect = sigl.Rect
  export type Ref<T> = sigl.VRef<T>
  export type RefSet<T extends $.Element<T>> = sigl.RefSet<T>
  export type RefItem<T> = sigl.RefItem<T>
  export type Scalar = sigl.Scalar
  export type Shape = sigl.Shape
  export type ShapeLike = sigl.ShapeLike
  export type StateInstance = sigl.StateInstance
  export type Super<T = any, U = object> = sigl.Super<T, U>
  export type Task = sigl.Task
}

export default $
