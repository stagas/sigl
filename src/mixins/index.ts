export * from './layout'
export * from './observed'
