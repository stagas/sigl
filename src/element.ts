import type {
  Class,
  CustomElementConstructor,
  EventHandler,
  StringKeys,
  StringOf,
  ValueConstructor,
} from 'everyday-types'
import { defineProperty, entries, getOwnProperty, omit, pick } from 'everyday-utils'
import { Hook, hook, render } from 'html-vdom'
import { Component, fromElement, Options } from 'html-vdom/from-element'
import { toFluent } from 'to-fluent'

import { applyAttrs, attrListener, AttrTypes } from './attrs'
import { ContextClass, EffectOptions, FluentFx, Fx, FxFn } from './context'
import { PropertySettings, protoPropertyMap } from './decorators'
import { Events } from './events'
import { RefProxy, Refs } from './refs'

import $ from '.'

export type { Component }

export type LifecycleEvents = {
  mounted: CustomEvent
  unmounted: CustomEvent
}

export interface JsxContext<T> {
  render: FluentFx<T, JSX.Element, void>
  part: FluentFx<T, JSX.Element, () => JSX.Element>
  ref: Refs<T>
}

export type ElementClass = CustomElementConstructor & {
  elementOptions?: Options<any>
  attributeKeys?: string[]
  propertyMap?: Map<string, PropertySettings>
}

export function element<T extends HTMLElement, I = HTMLElement>(
  Element: Class<T>,
  options?: Options<I>,
): Component<T, I>
export function element<I = HTMLElement>(options?: Options<I>): <T extends CustomElementConstructor>(superclass: T) => T
export function element(): <T extends CustomElementConstructor>(superclass: T) => T
export function element(ctorOrOptions?: Class<HTMLElement> | Options<any>, options?: Options<any>) {
  if (typeof ctorOrOptions === 'function') return fromElement(ctorOrOptions, options)
  if (typeof ctorOrOptions === 'object') options = ctorOrOptions
  return (superclass: any) => {
    const parent = superclass as ElementClass

    // get own prototype decorated properties map
    const ownPropertyMap = protoPropertyMap.get(superclass.prototype)! ?? new Map()

    // extract .attr keys
    const ownAttrKeys = [...ownPropertyMap].filter(([, settings]) => settings.attr).map(([key]) => key)

    // map from the lowercase (attribute) version to the actual property name
    const attrKeysMap = new Map(ownAttrKeys.map(x => [x.toLowerCase(), x]))

    // attribute types inferred/assigned at runtime
    const attrTypes = new Map<string, ValueConstructor>()

    // own observed attributes
    const ownObservedAttributes = [...attrKeysMap.keys()]

    //!? 'own observed attributes', ownObservedAttributes, superclass

    // @ts-ignore
    const ctor = class extends parent {
      static get elementOptions() {
        return options ?? super.elementOptions
      }

      static get observedAttributes() {
        return ownObservedAttributes.concat(super.observedAttributes ?? [])
      }

      static get attributeKeys() {
        return ownAttrKeys.concat(super.attributeKeys ?? [])
      }

      static get propertyMap() {
        return new Map([...ownPropertyMap, ...(super.propertyMap ?? [])])
      }

      static get exposedPropertyKeys() {
        return [...ctor.propertyMap]
          .filter(([, settings]) => settings.out)
          .map(([key]) => key) as StringKeys<InstanceType<typeof ctor>>[]
      }

      #isMain = true
      #attrData!: Pick<this, StringKeys<this>>
      #attrInitialUpdates: (() => void)[] = []

      $!: ContextClass<this> & JsxContext<this>
      host = this as unknown as HTMLElement & EventTarget & Events<this, LifecycleEvents>
      declare root?: ShadowRoot

      isMounted = false
      declare context: ContextClass<this> & JsxContext<this>
      declare onmounted: EventHandler<this, CustomEvent>
      declare onunmounted: EventHandler<this, CustomEvent>
      declare mounted?: ($: any) => void
      declare created?: ($: any) => void

      // goto://context.ts#ContextClass.createMemory
      declare _pendingProperties?: Map<symbol, [string, any[]]>

      declare _ctor: Class<any>

      constructor(...args: any[]) {
        super(...args)
        this.#isMain = new.target === ctor
        if (this.#isMain) {
          defineProperty.not.enumerable(this, '_ctor', ctor)
          this.#create()
          Object.assign(this, args[0])
          queueMicrotask(() => {
            this.#flushAttributeUpdates()
          })
        }
      }

      toJSON() {
        return pick(this, ctor.exposedPropertyKeys)
      }

      // goto://context.ts#ContextClass.createMemory
      _scheduleProperty(key: string, ...args: any[]) {
        const symbol = Symbol()
        this._pendingProperties ??= new Map()
        this._pendingProperties.set(symbol, [key, args])
        return symbol
      }

      _getPropertySettings(key: string) {
        return ctor.propertyMap.get(key)
      }

      #applyAttributes() {
        const attrs = pick(this, ctor.attributeKeys as StringKeys<this>[])

        for (const [key, value] of entries(attrs)) {
          const valueCtor = (value as any)?.constructor
          // TODO: maybe turn all as a .is constructor? Boolean needs special case
          if (valueCtor?.from) {
            const settings = ctor.propertyMap.get(key)!
            settings.is = valueCtor
          }
          //!? 'valueCtor', valueCtor.name
          //!? 'value', value
          const type = value == null
            ? AttrTypes.get(String)
            : AttrTypes.get(value as unknown as ValueConstructor)
              ?? (valueCtor
                ? AttrTypes.get(valueCtor as ValueConstructor)
                  /**
                   * if the inputs own class prototype implements toString
                   * then we can use that as the string attribute as
                   * the middleware will perform a .toString()
                   * goto://attrs.ts#AttrTypes
                   */ ?? (Object.hasOwn(valueCtor.prototype, 'toString')
                    ? AttrTypes.get(String)
                    : null)
                : null)
          if (!type) {
            throw new TypeError(
              `Attribute "${key}" is not valid type, must be either: String, Number, Boolean, null, undefined`
            )
          }
          //!? 'type', key, type.toString()
          attrTypes.set(key, type as ValueConstructor)
          attrKeysMap.set(key.toLowerCase(), key)
        }

        //!? 'apply attrs', attrs
        this.#attrData = attrs
        applyAttrs(this, attrs, this.#attrInitialUpdates)
      }

      #flushAttributeUpdates() {
        this.#attrInitialUpdates.splice(0).forEach(fn => fn())
      }

      #attachReactiveContext() {
        ContextClass.attach(this as any, omit($, ['transition']))
        this.host.on = ($.on as any).bind(this, this)
        this.host.dispatch = $.dispatchBind(this) as any
      }

      #registerChangeEvents() {
        const keys = new Set(ctor.exposedPropertyKeys)
        this.$.effect.keys(keys).task(() => {
          $.dispatch.composed(this.host, 'change')
        })
      }

      #create() {
        //!? 'creating', this

        // attach reactive context
        this.#attachReactiveContext()

        // attributes
        this.#applyAttributes()

        // listen for lifecycle events
        $.on(this.host).mounted(() => {
          //!? 'mounted', this

          // we flush here as well as at the queue level
          // because it can be created and mounted before
          // the next tick, so it will be allowed to create
          // attributes at this point.
          this.#flushAttributeUpdates()

          // register 'change' events from @$.out() and @$.attr.out()
          this.#registerChangeEvents()
        })

        $.on(this.host).unmounted(() => {
          //!? 'unmounted', this
          this.$.cleanup()
        })

        // jsx render helpers

        this.$.render = this.$.effect.cb((result: any) => {
          render(result, this.root ?? (this.root = $.shadow(this)))
          //!? 'render', this, result
          return false
        })

        this.$.part = toFluent(
          EffectOptions<this>,
          options =>
            (fn: FxFn<this, JSX.Element>, output?: JSX.Element) => {
              let update: Hook
              const cb = (value: JSX.Element) => {
                output = value
                //!? 'part', this, output
                if (update) queueMicrotask(update)
                return false
              }
              const Fn = () => {
                // lazily create effect when first used
                if (!update)
                  this.$.register(
                    new Fx({ fn, cb, options, dispose: cb })
                  )
                update = hook
                return output
              }
              return Fn
            }
        )

        this.$.ref = RefProxy(this.$)

        this.created?.(this.$)
      }

      // based on: https://stackoverflow.com/a/49773201/175416
      dispatchEvent(this: any, event: Event) {
        const onEventType = `on${event.type}`

        let pass = true

        // run inline listeners
        if (this.#isMain && Object.hasOwn(this, onEventType)) {
          let fn = this[onEventType]

          if (fn) {
            if (!fn) fn = attrListener(this.getAttribute(onEventType)!)
            pass = fn.call(this, event)
          }
        }

        // run prototype on<event>(event) listeners
        if (pass !== false) {
          const fn = getOwnProperty(super.constructor.prototype, onEventType)
          if (fn) pass = fn.call(this, event)
        }

        // run prototype <event>(context, event) listeners
        if (pass !== false) {
          const fn = getOwnProperty(super.constructor.prototype, event.type)
          if (fn) pass = fn.call(this, this.$, event)
        }

        // propagate event if pass is remains true
        if (pass !== false) super.dispatchEvent(event)

        return pass
      }

      connectedCallback() {
        if (!this.#isMain) {
          super.connectedCallback?.()
          return
        }

        //!? 'connected'
        //!dir this
        super.connectedCallback?.()
        if (!this.isMounted) {
          this.isMounted = true
          this.host.dispatch('mounted')
        }
      }

      disconnectedCallback() {
        if (!this.#isMain) {
          super.disconnectedCallback?.()
          return
        }

        super.disconnectedCallback?.()
        // When the element is moved, rather than removed,
        // disconnect and connect events fire together.
        // By observing the state at the end of the task
        // we only fire unmounted when it's been actually
        // removed from the dom and not reconnected
        // immediately.
        queueMicrotask(() => {
          if (!this.isConnected) {
            this.isMounted = false
            this.host.dispatch('unmounted')
          }
        })
      }

      propertyChangedCallback?(
        name: keyof this,
        oldValue: this[keyof this] | null,
        newValue: this[keyof this] | null,
      ): void

      attributeChangedCallback(name: StringOf<keyof this>, oldValue: string | null, newValue: string | null) {
        if (!this.#isMain) {
          super.attributeChangedCallback?.(name, oldValue, newValue)
          return
        }

        //!warn 'attributeChangedCallback:', name, newValue
        let key: StringOf<keyof this>
        if (key = attrKeysMap.get(name) as StringOf<keyof this>) {
          const { is } = ctor.propertyMap.get(key)!
          const prev = this.#attrData[key]
          const next = attrTypes.get(key)!(newValue) as unknown as this[StringOf<keyof this>]

          // type casting comparison for objects that serialize their attribute value
          const isDifferent = is ? prev != next : !Object.is(prev, next)

          //!? 'is different:', isDifferent, name, prev, next
          if (isDifferent) {
            this.#attrData[key] = next

            this[key] = is ? this[key] == next ? this[key] : is.from(next) : next

            //!? 'propertyChangedCallback:', key, next
            this.propertyChangedCallback?.(key, prev, next as unknown as this[StringOf<keyof this>])
          }
        }

        super.attributeChangedCallback?.(name, oldValue, newValue)
      }
    }

    // name the anonymous class with the decorated class' original name
    defineProperty(ctor, 'name', superclass.name)

    return ctor
  }
}
