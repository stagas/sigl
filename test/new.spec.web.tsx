/** @env browser */

/** @jsxImportSource ../src */

import { tick } from 'everyday-utils'
import { serialize } from 'serialize-whatever'
import $ from '../src'

let id = 0

describe('sigl', () => {
  describe('attrs', () => {
    it('cast number', () => {
      interface Foo extends $.Element<Foo> { }
      @$.element()
      class Foo extends HTMLElement {
        @$.attr() another = 123
      }
      customElements.define('x-foo' + id++, Foo)
      const foo = new Foo()
      expect(foo.another).toBe(123)
      foo.setAttribute('another', 'xyz')
      expect(foo.another).toBe(NaN)
    })

    it('delayed boolean', () => {
      interface Foo extends $.Element<Foo> { }
      @$.element()
      class Foo extends HTMLElement {
        @$.attr() flag = $.Boolean
      }
      customElements.define('x-foo' + id++, Foo)
      const foo = new Foo()
      expect(foo.flag).toBeUndefined()
      // foo.setAttribute('another', 'xyz')
      // expect(foo.another).toBe(NaN)
    })
  })

  it('works', () => {
    interface Foo extends $.Element<Foo, FooEvents> { }

    interface FooEvents {
      jump: CustomEvent<{ height: number }>
    }

    @$.element()
    class Foo extends HTMLElement {
      @$.attr() color = 'blue'
      @$.attr() another = 123
      @$.attr() withCapital = true
      @$.attr() notYet = $.Boolean

      button?: HTMLButtonElement
      result?: number = 123

      root = $.shadow(this)

      mounted() {
        const { $ } = this

        this.result = 5353

        $.effect(({ host }) => {
          host.onjump = ({ detail: { height } }) => {
            console.log(height)
          }
          host.on('jump')(({ detail: { height } }) => {
            console.log('height', height)
          })
          host.dispatch('jump', { height: 123 })
        })

        $.effect(({ notYet }) => {
          if (notYet) {
            // console.log('notYet:', notYet)
          }
        })

        $.render(({ color }) => (
          <div>
            {color}
            <button ref={$.ref.button}>click me</button>
          </div>
        ))
      }
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo()
    // foo.onmounted = () => {}
    // foo.host.onmounted = () => {}
    // foo.onjump
    // foo.host.onjump
    // foo.button
    // foo.color

    foo.setAttribute('color', 'red')
    expect(foo.getAttribute('color')).toBe('red')
    expect(foo.color).toBe('red')

    foo.setAttribute('another', 'red')
    expect(foo.getAttribute('another')).toBe('red') // TODO: this should be NaN but attrs aren't active yet
    expect(foo.another).toBe(NaN)
    foo.setAttribute('another', '456')
    expect(foo.getAttribute('another')).toBe('456')
    expect(foo.another).toBe(456)

    foo.toggleAttribute('withCapital')
    expect(foo.hasAttribute('withCapital')).toBe(true)
    expect(foo.withCapital).toBe(true)

    foo.toggleAttribute('withcapital')
    expect(foo.hasAttribute('withcapital')).toBe(false)
    expect(foo.withCapital).toBe(false)

    expect(foo.root.innerHTML).toBe('')
    document.body.appendChild(foo)
    expect(foo.root.innerHTML).toBe('<div>red<button>click me</button></div>')

    // foo.toggleAttribute('withcapital')
    expect(foo.hasAttribute('notyet')).toBe(false)
    expect(foo.notYet).toBeUndefined()
    foo.toggleAttribute('notyet')
    expect(foo.hasAttribute('notyet')).toBe(true)
    expect(foo.notYet).toBe(true)
    foo.toggleAttribute('notyet')
    expect(foo.hasAttribute('notyet')).toBe(false)
    expect(foo.notYet).toBe(false)
  })

  it('inheritance works', () => {
    const called: number[] = []
    let calledIndex = 0

    // const ctorCalled: number[] = []
    // let ctorIndex = 0

    let listenerCalled = 0

    interface Bar extends $.Element<Bar> { }

    @$.element()
    class Bar extends HTMLElement {
      @$.attr() parentObserved = true
      // constructor(is) {
      //   super(true)
      //   console.log(is)
      //   ctorCalled.push(ctorIndex++)
      // }
      mounted($: this['$']) {
        called.push(calledIndex++)
        $.effect(({ parentObserved: _ }) => {
          parentSeen++
        })
      }
    }

    interface Foo extends $.Element<Foo> { }

    let seen = 0

    let parentSeen = 0
    let childSeen = 0

    @$.element()
    class Foo extends $.inherit(Bar) {
      observed = true
      // constructor(is) {
      //   super(true)
      //   console.log(is)
      //   ctorCalled.push(ctorIndex++)
      // }
      mounted($: this['$']) {
        called.push(calledIndex++)
        $.effect(({ observed: _ }) => {
          seen++
        })

        $.effect(({ parentObserved: _ }) => {
          childSeen++
        })
      }
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo()
    foo.onmounted = () => listenerCalled++
    document.body.appendChild(foo)

    expect(called).toEqual([0, 1])
    // expect(ctorCalled).toEqual([0, 1])
    expect(listenerCalled).toBe(1)
    expect(seen).toBe(1)
    expect(parentSeen).toBe(1)
    expect(childSeen).toBe(1)

    customElements.define('x-foo' + id++, Bar)
    const bar = new Bar()
    bar.onmounted = () => listenerCalled++
    document.body.appendChild(bar)

    expect(called).toEqual([0, 1, 2])
    expect(listenerCalled).toBe(2)
    expect(seen).toBe(1)
    expect(parentSeen).toBe(2)
    expect(childSeen).toBe(1)
  })

  it('dispatching events', () => {
    const called: string[] = []
    @$.element()
    class Zoo extends HTMLElement {
      // @ts-ignore
      onclick() {
        called.push('zoo-click')
      }
      onyo() {
        called.push('zoo-yo')
      }
      mounted() {
        called.push('zoo-mounted')
      }
    }

    @$.element()
    class Bar extends Zoo {
      onyo() {
        called.push('bar-yo')
      }
      mounted() {
        called.push('bar-mounted')
      }
    }

    @$.element()
    class Foo extends Bar {
      onyo() {
        called.push('foo-yo')
      }
      onclick() {
        called.push('foo-click')
      }
      mounted() {
        called.push('foo-mounted')
      }
    }

    customElements.define('x-foo' + id++, Foo)

    const foo = new Foo()
    foo.onclick = () => {
      called.push('inline-click')
    }
    foo.onyo = () => {
      called.push('inline-yo')
    }
    foo.dispatchEvent(new MouseEvent('click'))
    foo.dispatchEvent(new CustomEvent('yo'))
    expect(called).toEqual([
      'inline-click',
      'foo-click',
      'zoo-click',
      'inline-yo',
      'foo-yo',
      'bar-yo',
      'zoo-yo',
    ])
    document.body.appendChild(foo)
    expect(called).toEqual([
      'inline-click',
      'foo-click',
      'zoo-click',
      'inline-yo',
      'foo-yo',
      'bar-yo',
      'zoo-yo',
      'foo-mounted',
      'bar-mounted',
      'zoo-mounted',
    ])
  })

  it('prop reducers', async () => {
    interface Foo extends $.Element<Foo> { }

    @$.element()
    class Foo extends HTMLElement {
      min = 30
      max = 100
      value = 50
      scale = $(this).reduce(({ min, max }) => max - min)
      normal = $(this).reduce(({ min, scale, value }) => (value - min) / scale)

      root = $.shadow(this)

      mounted($: this['$']) {
        $.render(({ normal }) => <div>{normal}</div>)
      }
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo()
    document.body.appendChild(foo)
    expect(foo.root.innerHTML).toMatchSnapshot()
  })

  it('mixins', async () => {
    interface Foo extends $.Element<Foo> { }

    const order: string[] = []

    const runnable = $.mixin(superclass =>
      class extends superclass {
        speed = 10
        mounted() {
          order.push('runnable')
        }
      }
    )

    const swimmable = $.mixin(superclass =>
      class extends superclass {
        depth = 10
        mounted($: $.Element<this>['$']) {
          $.depth
          order.push('swimmable')
        }
      }
    )

    @$.element()
    class Foo extends $.mix(HTMLElement, runnable, swimmable) {
      root = $.shadow(this)
      mounted($: this['$']) {
        order.push('foo')
        $.render(({ speed, depth }) => <div>{speed * depth}</div>)
      }
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo()
    document.body.appendChild(foo)
    expect(foo.root.innerHTML).toMatchSnapshot()
    expect(order).toMatchSnapshot()
  })

  it('constructor state argument', async () => {
    interface Foo extends $.Element<Foo> { }

    @$.element()
    class Foo extends $.inherit(HTMLElement) {
      value = 123
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo({ value: 456 })
    document.body.appendChild(foo)
    expect(foo.value).toBe(456)
  })

  it('refs', async () => {
    interface Foo extends $.Element<Foo> { }

    @$.element()
    class Foo extends HTMLElement {
      button?: HTMLButtonElement
      mounted($: this['$']) {
        $.render(() => <button ref={$.ref.button}>press</button>)
      }
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo()
    document.body.appendChild(foo)
    await tick()
    expect(foo.button).toBeInstanceOf(HTMLButtonElement)
  })

  it('ref map', async () => {
    interface ButtonElement extends $.Element<ButtonElement> { }
    @$.element()
    class ButtonElement extends $.inherit(HTMLElement) { }
    const Button = $.element(ButtonElement)

    interface Foo extends $.Element<Foo> { }

    let count = 0

    @$.element()
    class Foo extends HTMLElement {
      buttons = new $.RefMap<ButtonElement>([
        ['foo', {}],
        ['bar', {}],
      ])
      mounted($: this['$']) {
        $.effect(({ buttons: _ }) => {
          count++
        })
        $.render(({ buttons }) => buttons.map(item => <Button key={item.ref.key} {...item}>press</Button>))
      }
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo()
    document.body.appendChild(foo)
    expect(count).toBe(1)
    await tick()
    expect(foo.buttons.get('foo')!.ref.current).toBeInstanceOf(ButtonElement)
    expect(foo.buttons.get('bar')!.ref.current).toBeInstanceOf(ButtonElement)
    expect(count).toBe(3)
  })

  // it('ref map recover', async () => {
  //   interface BarElement extends $.Element<BarElement> {}
  //   @$.element()
  //   class BarElement extends $(HTMLElement) {
  //     @$.attr() value = 0
  //     root = this
  //     mounted($: this['$']) {
  //       $.render(({ value }) => <span>{value}</span>)
  //     }
  //   }

  //   let count = 0

  //   interface Foo extends $.Element<Foo> {}
  //   @$.element()
  //   class Foo extends HTMLElement {
  //     Bar = $.element(BarElement)

  //     bars = [
  //       { key: 'foo', current: new BarElement({ value: 123 }) },
  //       { key: 'bar', current: new BarElement({ value: 456 }) },
  //     ]

  //     root = this

  //     mounted($: this['$']) {
  //       $.effect(({ bars: _ }) => {
  //         count++
  //       })
  //       $.render(({ Bar, bars }) => (
  //         bars.map(ref => <Bar key={ref.key} ref={ref} />)
  //       ))
  //     }
  //   }

  //   customElements.define('x-foo' + id++, Foo)
  //   const foo = new Foo()
  //   document.body.appendChild(foo)
  //   expect(count).toBe(1)
  //   await tick()
  //   // expect(foo.bars.get('foo')!.current).toBeInstanceOf(BarElement)
  //   // expect(foo.bars.get('bar')!.current).toBeInstanceOf(BarElement)
  //   expect(count).toBe(1)
  //   expect(foo.innerHTML).toMatchSnapshot()
  // })

  it('serializing', async () => {
    interface Foo extends $.Element<Foo> { }

    @$.element()
    class Foo extends $.inherit(HTMLElement) {
      @$.out() value = 123
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo({ value: 456 })
    document.body.appendChild(foo)
    expect(foo.value).toBe(456)
    expect(foo.toJSON()).toEqual({
      value: 456,
    })
  })

  it('serializing ref map', async () => {
    interface BarElement extends $.Element<BarElement> { }
    @$.element()
    class BarElement extends $.inherit(HTMLElement) {
      @$.attr.out() value = 0
      root = this
      mounted($: this['$']) {
        $.render(({ value }) => <span>{value}</span>)
      }
    }

    let count = 0

    interface Foo extends $.Element<Foo> { }
    @$.element()
    class Foo extends HTMLElement {
      Bar = $.element(BarElement)

      @$.out() bars = new $.RefMap<BarElement>([
        ['foo', { value: 123 }],
        ['bar', { value: 456 }],
      ])

      root = this

      mounted($: this['$']) {
        $.effect(({ bars: _ }) => {
          count++
        })
        $.render(({ Bar, bars }) => (
          bars.map(item => <Bar {...item} />)
        ))
      }
    }

    customElements.define('x-foo' + id++, Foo)
    const foo = new Foo()
    document.body.appendChild(foo)
    expect(count).toBe(1)
    await tick()
    // expect(foo.bars.get('foo')!.current).toBeInstanceOf(BarElement)
    // expect(foo.bars.get('bar')!.current).toBeInstanceOf(BarElement)
    expect(count).toBe(3)
    // expect(foo.innerHTML).toMatchSnapshot()
    expect(serialize(foo, 2)).toMatchSnapshot()
  })

  it('state', () => {
    type FooEvents = {
      statechange: CustomEvent<$.StateInstance>
      // handlestart: CustomEvent
      handleend: CustomEvent
    }

    const FooState = {
      Idle: 'idle',
      Handle: 'handle',
    } as const

    interface FooElement extends $.Element<FooElement, FooEvents> { }

    let mountedState: any

    @$.element()
    class FooElement extends HTMLElement {
      @$.attr() state = $(this).state(FooState)

      mounted($: FooElement['$']) {
        $.effect(({ state }) => {
          mountedState = state
        })
      }
      //   $.effect(({ state }) => {
      //     $.state.push(FooState.Handle)
      //     $.state.pop(FooState.Handle)
      //     state.cancel(FooState.Handle)
      //     // const x = state.current
      //     $.on(state).handlestart(() => {
      //       //
      //     })
      //   })
      // }
    }

    customElements.define('x-foo' + id++, FooElement)
    const foo = new FooElement()
    document.body.appendChild(foo)
    // expect(foo.state).toBe(FooState.Idle)
    expect(foo.$.state).toBeInstanceOf($.State)
    expect(mountedState).toBeInstanceOf($.State)
    // TODO: assert
  })
})
