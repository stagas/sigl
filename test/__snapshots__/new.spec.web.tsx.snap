
exports[`sigl prop reducers 1`] = `"<div>0.2857142857142857</div>"`;

exports[`sigl mixins 1`] = `"<div>100</div>"`;

exports[`sigl mixins 2`] = `
Array [
  "foo",
  "swimmable",
  "runnable",
]
`;

exports[`sigl serializing ref map 1`] = `
"{
  \\"$r\\": 1,
  \\"$t\\": \\"Foo\\",
  \\"$v\\": {
    \\"bars\\": {
      \\"$r\\": 2,
      \\"$t\\": \\"RefMap\\",
      \\"$v\\": [
        {
          \\"$r\\": 3,
          \\"$t\\": \\"Array\\",
          \\"$v\\": [
            \\"foo\\",
            {
              \\"$r\\": 4,
              \\"$t\\": \\"Object\\",
              \\"$v\\": {
                \\"value\\": 123
              }
            }
          ]
        },
        {
          \\"$r\\": 5,
          \\"$t\\": \\"Array\\",
          \\"$v\\": [
            \\"bar\\",
            {
              \\"$r\\": 6,
              \\"$t\\": \\"Object\\",
              \\"$v\\": {
                \\"value\\": 456
              }
            }
          ]
        }
      ]
    }
  }
}"
`;
